Stretched Sprite Origin Fix

RobertBColton opened the pull request:
This is just a small fix to address #1099 which I have confirmed is in fact a bug. Mark Overmars is very explicit in the manual that `draw_sprite_stretched` "fills" the area as opposed to `draw_sprite` which places the origin at the position specified.

This simple fix just removes the offset of the primitive drawn by the sprite's origin. I've additionally added a test to the drawing test so we know if somebody tries to change it again. Although I do not actually change draw_sprite_stretched_ext here, it is still affected because it's just a wrapper that calls draw_sprite_stretched with its color and alpha.

In my opinion, GameMaker is correct about this behavior as the sprite origin is basically useless when drawing the sprite stretched over an area. I feel this is especially true considering the sprite origin is not scaled with the rest of the sprite, so it's better to just not use it both for compatibility and because it's saner.

### References
The GameMaker 6 through 8 manuals state the following:
>draw_sprite(sprite,subimg,x,y) Draws subimage subimg (-1 = current) of the sprite with its **origin at position** (x,y). (Without color blending and no alpha transparency.)
draw_sprite_stretched(sprite,subimg,x,y,w,h) Draws the sprite stretched so that it **fills the region** with top-left corner (x,y) and width w and height h.

The new GameMaker: Studio manual is even more explicit:

https://docs2.yoyogames.com/source/_build/3_scripting/4_gml_reference/drawing/sprites_and_tiles/draw_sprite_stretched.html
>NOTE: When drawing with this function, the sprite x offset and y offset are ignored and the sprite will be drawn with the top left corner at the specified x / y position in the room.

https://docs2.yoyogames.com/source/_build/3_scripting/4_gml_reference/drawing/sprites_and_tiles/draw_sprite_stretched_ext.html
>This function ignores the sprite's origin.

### ENIGMA Master
![master Stretched Sprite Origin](https://user-images.githubusercontent.com/3212801/48814491-f0300400-ed08-11e8-82b3-07550a373ddb.png)
### This pull request
![pr Stretched Sprite Origin](https://user-images.githubusercontent.com/3212801/48814502-fde58980-ed08-11e8-97ae-a1a978873acf.png)
### GameMaker 8
![GM8 Stretched Sprite Origin](https://user-images.githubusercontent.com/3212801/48814478-e27a7e80-ed08-11e8-9ff8-b2af621c0bc1.png)
### GameMaker: Studio
![GMS Stretched Sprite Origin](https://user-images.githubusercontent.com/3212801/48814457-cbd42780-ed08-11e8-9105-1f3861cb79d7.png)

RobertBColton commented:
The bot is correct that changes have been introduced to the drawing test, but this time, they are intentional.

JoshDreamland commented on CommandLine/testing/Tests/draw_test.sog/draw.edl:
```diff
@@ -40,3 +40,7 @@ draw_set_color(c_white);
 draw_batch_flush();
+
+// stretched sprite should ignore origin
+draw_sprite_stretched(spr,0,0,0,25,25);
```
There are *soooo* many sprite tests to add. Can you extract these to a new sprites test?

RobertBColton replied:
Not yet, because that automatically adds about a 1-3 minute overhead. Wait until we're actually ready, I have a feeling I may need to fix some more anomalies, and this isn't hard to move at all.

RobertBColton commented:
Alright, so I performed some additional testing on GameMaker 5 because of what I noticed in my comment on #1099. It seems Mark Overmars scaled the sprite origin in GM5 as you can see in the following screen. I will add a compatibility fix for that.
![GM5 Stretched Sprite Origin](https://user-images.githubusercontent.com/3212801/48817481-b1a04680-ed14-11e8-9e04-856dd4c9c973.png)

RobertBColton commented:
Alright: f9e2155a594b22f0eb2e1d7c98721e99f06d79df

I've add a compliance setting for the behavior now for GameMaker 5 projects. I had to make several adjustments to the way `GM_COMPATIBILITY_VERSION` is implemented though.
* Had to include "Preprocessor_Environment_Editable/GAME_SETTINGS.h" for `GM_COMPATIBILITY_VERSION` to be checked in `GSsprite.cpp`.
* Had to comment out the declaration of `void ABORT_ON_ALL_ERRORS()` to stop multiple definition errors, which is fine because it wasn't actually being used anyway.
* I included `GAME_SETTINGS.h` in `PFwindow.h` so that I could check the `GM_COMPATIBILITY_VERSION` for `window_handle()`. Apparently `SHELLmain.cpp` is including `PFwindow.h` at some time before it, although I noticed `API_Switchboard.h` is in fact including `PFwindow.h` after `GAME_SETTINGS.h` in `SHELLmain.cpp`.

RobertBColton commented:
Alright, I've added a compliance setting for the behavior now for GameMaker 5 projects. I had to make several adjustments to the way `GM_COMPATIBILITY_VERSION` is implemented though.
* Had to include "Preprocessor_Environment_Editable/GAME_SETTINGS.h" for `GM_COMPATIBILITY_VERSION` to be checked in `GSsprite.cpp`.
* Had to change the declaration of `void ABORT_ON_ALL_ERRORS()` to stop multiple definition errors. I changed it to just define the value and then check it in the engine inside `abort_on_errors`.
* I included `GAME_SETTINGS.h` in `PFwindow.h` so that I could check the `GM_COMPATIBILITY_VERSION` for `window_handle()`. Apparently `SHELLmain.cpp` is including `PFwindow.h` at some time before it, although I noticed `API_Switchboard.h` is in fact including `PFwindow.h` after `GAME_SETTINGS.h` in `SHELLmain.cpp`.

RobertBColton commented:
Alright: f9e2155a594b22f0eb2e1d7c98721e99f06d79df

I've added a compliance setting for the behavior now for GameMaker 5 projects. I had to make several adjustments to the way `GM_COMPATIBILITY_VERSION` is implemented though.
* Had to include "Preprocessor_Environment_Editable/GAME_SETTINGS.h" for `GM_COMPATIBILITY_VERSION` to be checked in `GSsprite.cpp`.
* Had to comment out the declaration of `void ABORT_ON_ALL_ERRORS()` to stop multiple definition errors, which is fine because it wasn't actually being used anyway.
* I included `GAME_SETTINGS.h` in `PFwindow.h` so that I could check the `GM_COMPATIBILITY_VERSION` for `window_handle()`. Apparently `SHELLmain.cpp` is including `PFwindow.h` at some time before it, although I noticed `API_Switchboard.h` is in fact including `PFwindow.h` after `GAME_SETTINGS.h` in `SHELLmain.cpp`.

RobertBColton commented:
Alright, I've added a compliance setting for the behavior now for GameMaker 5 projects. I had to make several adjustments to the way `GM_COMPATIBILITY_VERSION` is implemented though.
* Had to include "Preprocessor_Environment_Editable/GAME_SETTINGS.h" for `GM_COMPATIBILITY_VERSION` to be checked in `GSsprite.cpp`.
* Had to change the macro `void ABORT_ON_ALL_ERRORS()` to stop multiple definition errors. I changed it to just define the value and then check it in the engine inside `abort_on_errors`.
* I included `GAME_SETTINGS.h` in `PFwindow.h` so that I could check the `GM_COMPATIBILITY_VERSION` for `window_handle()`. Apparently `SHELLmain.cpp` is including `PFwindow.h` at some time before it, although I noticed `API_Switchboard.h` is in fact including `PFwindow.h` after `GAME_SETTINGS.h` in `SHELLmain.cpp`.

JoshDreamland commented:
Including that header will cause files to be rebuilt each time that header changes. Depending on what's in that file, that is probably every time. This change, then, would vastly slow the build. Don't do less-than checks on the compatibility version without first checking that it is `defined()`, and otherwise, don't include that header.
